src/coreclr/src/jit/morph.cpp

```diff
@@ -18370,6 +18382,108 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+    //    user - the indirection's user node
+    //
+    void MorphLocalIndir(const Value& val, GenTree* user)
+    {
+        assert(val.IsLocation());
+
+        GenTree* indir = val.Node();
+        assert(indir->OperIs(GT_IND, GT_FIELD));
+
+        if ((indir->OperIs(GT_IND) && indir->AsIndir()->IsVolatile()) ||
+            (indir->OperIs(GT_FIELD) && indir->AsField()->IsVolatile()))
+        {
+            return;
+        }
+
+        if (varTypeIsStruct(indir->TypeGet()))
+        {
+            // TODO-ADDR: Skip struct indirs for now, they require struct typed LCL_FLDs.
+            // Also skip SIMD indirs for now, SIMD typed LCL_FLDs works most of the time
+            // but there are exceptions - fgMorphFieldAssignToSIMDIntrinsicSet for example.
+            return;
+        }
+
+        LclVarDsc* varDsc = m_compiler->lvaGetDesc(val.LclNum());
+
+        if (varDsc->TypeGet() != TYP_STRUCT)
+        {
+            // TODO-ADDR: Skip integral/floating point variables for now, they're more
+            // complicated to transform. We can always turn an indirect access of such
+            // a variable into a LCL_FLD but that blocks enregistration so we need to
+            // detect those case where we can use LCL_VAR instead, perhaps in conjuction
+            // with CAST and/or BITCAST.
+            // Also skip SIMD variables for now, fgMorphFieldAssignToSIMDIntrinsicSet and
+            // others need to be updated to recognize LCL_FLDs.
+            return;
+        }
+
+        if (varDsc->lvPromoted || varDsc->lvIsStructField || m_compiler->lvaIsImplicitByRefLocal(val.LclNum()))
+        {
+            // TODO-ADDR: For now we ignore promoted and "implict by ref" variables,
+            // they require additional changes in subsequent phases
+            // (e.g. fgMorphImplicitByRefArgs does not handle LCL_FLD nodes).
+            return;
+        }
+
+        if (val.Offset() > UINT16_MAX)
+        {
+            // TODO-ADDR: We should morph the address of the indir using MorphLocalAddress,
+            // once the current usage restrictions on LCL_VAR|FLD_ADDR are lifted.
+            return;
+        }
+
+        FieldSeqNode* fieldSeq = val.FieldSeq();
+
+        if (indir->OperIs(GT_IND) && (fieldSeq != nullptr) && (fieldSeq != FieldSeqStore::NotAField()))
+        {
+            // If we have an IND node and a field sequence then they should have the same type.
+            // Otherwise it's best to forget the field sequence since the resulting LCL_FLD
+            // doesn't match a real struct field. Value numbering protects itself from such
+            // mismatches but there doesn't seem to be any good reason to generate a LCL_FLD
+            // with a mismatched field sequence only to have to ignore it later.
+
+            assert(!fieldSeq->IsPseudoField());
+
+            if (indir->TypeGet() != JITtype2varType(m_compiler->info.compCompHnd->getFieldType(fieldSeq->m_fieldHnd)))
```

Comment:
`fieldSeq->m_fieldHnd` is incorrect, we need the handle of the last field in the sequence - i.e., if we have `a.b.c` then we need the type of `c`.